tools/pack_glb.py

```diff
@@ -0,0 +1,202 @@
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import json
+import logging
+import os
+from pathlib import Path
+import sys
+from typing import Any, BinaryIO
+from typing_extensions import NotRequired, TypedDict
+from urllib.parse import urlparse
+
+
+def align_to_4(value: int) -> int:
+    return (value + 3) & ~3
+
+
+class GLTFError(Exception):
+    """General exception class for GLTF and GLB errors"""
+
+
+class GLTFBufferDescription(TypedDict):
+    byteLength: int
+    uri: NotRequired[str]
+
+
+class GLTFBufferViewDescription(TypedDict):
+    buffer: int
+    byteLength: int
+    byteOffset: int
+
+
+class GLTFImageDescription(TypedDict):
+    uri: NotRequired[str]
+    bufferView: int
+    mimeType: str
+
+
+class GLTFDescription(TypedDict):
+    buffers: list[GLTFBufferDescription]
+    bufferViews: list[GLTFBufferViewDescription]
+    images: list[GLTFImageDescription]
+
+
+class GLB:
+    """Holds the data necessary to create a GLB file."""
+
+    def __init__(self, description: GLTFDescription, source_dir: Path) -> None:
+        self.description = description
+        self.source_dir = source_dir
+        self.bin_length = 0
+        self.source_files: list[Path] = []
+
+    def pack(self) -> None:
+        """Rewrite all references to external files as references to a packed buffer."""
+        self._pack_buffers()
+        self._pack_images()
+        self.description["buffers"] = [{"byteLength": self.bin_length}]
+
+        if self.bin_length > 2**32 - 1:
+            logging.warning(
+                "Packed internal buffer exceeds recommended max byte length of 2**32-1: %i",
+                self.bin_length,
+            )
+
+    def _pack_buffers(self) -> None:
+        # We are packing potentially multiple external binary buffer files into a single internal buffer
+        # This maps the index of each external buffer to an offset in the internal buffer
+        buffer_offsets: dict[int, int] = {}
+
+        for index, buffer in enumerate(self.description["buffers"]):
+            if "uri" in buffer:
+                # this doesn't handle data uris; we can add support if necessary
+                uri = urlparse(buffer["uri"])
+                if uri.scheme != "" or uri.netloc != "":
+                    raise GLTFError(f'support for uri not implemented: "{uri}"')
+                # record the path of the locally stored external buffer to be packed
+                self.source_files.append(Path(uri.path))
+                buffer_offsets[index] = self.bin_length
+            else:
+                if index != 0:
```

Comment:
Maybe add a comment to https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#glb-stored-buffer so we understand why the first buffer omits the property?

Reply:
Done